pkg/virt-controller/watch/vm.go

```diff
@@ -83,209 +126,427 @@ func (c *VMController) runWorker() {
+	needsSync := c.podExpectations.SatisfiedExpectations(key) && c.handoverExpectations.SatisfiedExpectations(key)
 
-		// Deep copy the object, so that we can safely manipulate it
-		vmCopy := vm.DeepCopy()
-		logger = log.Log.Object(vmCopy)
+	var syncErr error
+	if needsSync {
+		syncErr = c.sync(vm, pods)
+	}
+	return c.updateStatus(vm, pods, syncErr)
+}
 
-		// Check if there are already outdated VM Pods
-		pods, err := c.vmService.GetRunningVMPods(vmCopy)
-		if err != nil {
-			logger.Reason(err).Error("Fetching VM pods failed.")
-			return err
-		}
+func (c *VMController) updateStatus(vm *virtv1.VirtualMachine, pods []*k8sv1.Pod, syncErr error) error {
 
-		// If there are already pods, delete them before continuing ...
-		if len(pods.Items) > 0 {
-			logger.Error("VM Pods do already exist, will clean up before continuing.")
-			if err := c.vmService.DeleteVMPod(vmCopy); err != nil {
-				logger.Reason(err).Error("Deleting VM pods failed.")
-				return err
+	// Don't process states where the vm is clearly owned by virt-handler
+	if vm.IsRunning() || vm.IsScheduled() {
+		return nil
+	}
+
+	vmCopy := vm.DeepCopy()
+
+	if vm.IsUnprocessed() || vm.IsScheduling() {
+		if len(pods) > 0 {
+
+			containersAreReady := podIsReady(pods[0])
+
+			podIsTerminating := podIsDownOrGoingDown(pods[0])
+
+			// vm is still owned by the controller but pod is already handed over,
+			// so let's hand over the vm too
+			if c.isPodOwnedByHandler(pods[0]) {
+
+				vmCopy.Status.Interfaces = []virtv1.VirtualMachineNetworkInterface{
+					{
+						IP: pods[0].Status.PodIP,
+					},
+				}
+				vmCopy.Status.Phase = virtv1.Scheduled
+				if vmCopy.Labels == nil {
+					vmCopy.Labels = map[string]string{}
+				}
+				vmCopy.ObjectMeta.Labels[virtv1.NodeNameLabel] = pods[0].Spec.NodeName
+				vmCopy.Status.NodeName = pods[0].Spec.NodeName
+			} else if !podIsTerminating && !containersAreReady {
+				vmCopy.Status.Phase = virtv1.Scheduling
+			} else if podIsTerminating {
+				vmCopy.Status.Phase = virtv1.Failed
 			}
-			// the pod informer will reenqueue the key as a result of it being deleted.
-			return nil
+		} else if len(pods) == 0 && vm.IsScheduling() {
+			// someone other then the controller deleted the pod unexpectedly
+			vmCopy.Status.Phase = virtv1.Failed
+		}
+	} else if vm.IsFinal() {
+		if len(pods) == 0 {
+			controller.RemoveFinalizer(vmCopy, virtv1.VirtualMachineFinalizer)
 		}
+	}
 
-		// Defaulting and setting constants
-		// TODO move defaulting to virt-api
-		kubev1.SetObjectDefaults_VirtualMachine(vmCopy)
+	// Select the right failure reason in case we have an error
+	// TODO this reason detetion is not 100 % accurate
+	reason := ""
+	if len(pods) == 0 {
+		reason = "FailedCreate"
+	} else if len(pods) == 1 && vm.IsFinal() || vm.DeletionTimestamp != nil {
+		reason = "FailedDelete"
+	} else {
+		reason = "FailedHandOver"
+	}
+	controller.NewVirtualMachineConditionManager().CheckFailure(vmCopy, syncErr, reason)
```

Comment:
is it possible for a vm to recover from any of these errors?

Reply:
Yes. Normally the reason for these errors are transient connection errors. We simply re-enqueue on such errors with a backoff.